cranelift/object/src/backend.rs

```diff
@@ -264,22 +265,30 @@ impl Backend for ObjectBackend {
         }
 
         let symbol = self.data_objects[data_id].unwrap();
-        let section_kind = if let Init::Zeros { .. } = *init {
-            if tls {
-                StandardSection::UninitializedTls
+        let section = 
+            if datasection.is_none() {
+                let section_kind = if let Init::Zeros { .. } = *init {
+                    if tls {
+                        StandardSection::UninitializedTls
+                    } else {
+                        StandardSection::UninitializedData
+                    }
+                } else if tls {
+                    StandardSection::Tls
+                } else if writable {
+                    StandardSection::Data
+                } else if relocs.is_empty() {
+                    StandardSection::ReadOnlyData
+                } else {
+                    StandardSection::ReadOnlyDataWithRel
+                };
+                self.object.section_id(section_kind)
             } else {
-                StandardSection::UninitializedData
-            }
-        } else if tls {
-            StandardSection::Tls
-        } else if writable {
-            StandardSection::Data
-        } else if relocs.is_empty() {
-            StandardSection::ReadOnlyData
-        } else {
-            StandardSection::ReadOnlyDataWithRel
-        };
-        let section = self.object.section_id(section_kind);
+                let (seg, sec) = &datasection.as_ref().unwrap();
```

Comment:
Can you add an assertion that it isn't a tls data object? Those only work for `StandardSection::Tls`.